### ISSUE TITLE
-Wfloat-equal does not check for exact representation correctly

### ISSUE THREAD

**Issue by Halalaluyafail3** [llvm/llvm-project#220746]
The following program demonstrates the issue:
```cpp
bool test(double d){
    return d==1.100000000000000088817841970012523233890533447265625;
}
```
This generates -Wfloat-equal, but writing the equivalent value as a hexadecimal floating literal `0X1199999999999AP-52` instead of `1.100000000000000088817841970012523233890533447265625` causes no warning to be generated.

**Comment by anondeveg** [comment]
Iam currently working on this!

**Comment by anondeveg** [comment]
after some investigations 
the problem is in this if condition:
```cpp
  // Special case: check for comparisons against literals that can be exactly
  //  represented by APFloat.  In such cases, do not emit a warning.  This
  //  is a heuristic: often comparison against such literals are used to
  //  detect if a value in a variable has not changed.  This clearly can
  //  lead to false negatives.
  if (const auto *FLL = dyn_cast<FloatingLiteral>(LeftExprSansParen)) {
    if (FLL->isExact())
      return;
  } else if (const auto *FLR = dyn_cast<FloatingLiteral>(RightExprSansParen))
    if (FLR->isExact())
      return;
```

the problem is with the way isExact is calculated, with hex format it is more precise so it gets assigned to be exact thus returning without raising the error, while the decimal format dosen't get assigned as exact so it dosen't trigger the if condition thus raising the warning.

i believe that this is not under the expected false negatives mentioned in the comment, so its an actual bug in `isExact` if it is expected to give the same result for the same value despite it's form.

Especially since GCC raises a warning for the same case:
```bash
gcc -Wall -Wfloat-equal test.cpp

t.cpp: In function ‘bool test(double)’:
t.cpp:3:14: warning: comparing floating-point with ‘==’ or ‘!=’ is unsafe [-Wfloat-equal]
    3 |      return d==0X1199999999999AP-52;
      |             ~^~~~~~~~~~~~~~~~~~~~~~
```
will look further into this.

**Comment by shafik** [comment]
CC @jcranmer-intel @jhuber6

**Comment by anondeveg** [comment]
> Especially since GCC raises a warning for the same case:
> 
> gcc -Wall -Wfloat-equal test.cpp
> 
> t.cpp: In function ‘bool test(double)’:
> t.cpp:3:14: warning: comparing floating-point with ‘==’ or ‘!=’ is unsafe [-Wfloat-equal]
>     3 |      return d==0X1199999999999AP-52;
>       |             ~^~~~~~~~~~~~~~~~~~~~~~
> 
> will look further into this.

i tried gcc with `0.5f` and it did raise an error while clang didn't, it looks like gcc dosen't have exact-literal supression at all.

should we even raise a warning here or is it expected to be suppressed?

**Comment by Halalaluyafail3** [comment]
> i tried gcc with `0.5f` and it did raise an error while clang didn't, it looks like gcc dosen't have exact-literal supression at all.
> 
> should we even raise a warning here or is it expected to be suppressed?

GCC warns for any floating equality comparison, even those that make sense like zero, infinity, or comparing against the same value to test for NaN.
